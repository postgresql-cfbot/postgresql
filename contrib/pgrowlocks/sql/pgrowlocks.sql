--
-- Test pgrowlocks extension
--

CREATE EXTENSION pgrowlocks;

-- set up
CREATE TABLE test(c int);

-- check permission to use pgrowlocks extension;

CREATE ROLE regress_priv_user1 superuser login;
CREATE ROLE regress_priv_user2;
CREATE ROLE regress_priv_user3;

SET SESSION AUTHORIZATION regress_priv_user1;

SELECT locked_row, locker, multi, xids, modes FROM pgrowlocks('test'); -- ok

SET SESSION AUTHORIZATION regress_priv_user2;
SELECT locked_row, locker, multi, xids, modes FROM pgrowlocks('test'); -- fail

-- switch to superuser
\c -

GRANT SELECT ON test TO regress_priv_user3;

SET SESSION AUTHORIZATION regress_priv_user3;
SELECT locked_row, locker, multi, xids, modes FROM pgrowlocks('test'); -- ok

