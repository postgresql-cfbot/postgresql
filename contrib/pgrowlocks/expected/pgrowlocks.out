unused step name: s2_tuplock1
unused step name: s2_tuplock2
unused step name: s2_tuplock3
unused step name: s2_tuplock4
Parsed test spec with 2 sessions

starting permutation: s1_begin s1_lcksvpt s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes              
----------+-----+-------------------
(0,1)     |t    |{"Key Share",Share}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes              
----------+-----+-------------------
(0,1)     |t    |{"Key Share",Share}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes              
----------+-----+-------------------
(0,1)     |t    |{"Key Share",Share}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes              
----------+-----+-------------------
(0,1)     |t    |{"Key Share",Share}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                            
----------+-----+---------------------------------
(0,1)     |t    |{"Key Share","For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                            
----------+-----+---------------------------------
(0,1)     |t    |{"Key Share","For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                            
----------+-----+---------------------------------
(0,1)     |t    |{"Key Share","For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                            
----------+-----+---------------------------------
(0,1)     |t    |{"Key Share","For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                     
----------+-----+--------------------------
(0,1)     |t    |{"Key Share","For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                     
----------+-----+--------------------------
(0,1)     |t    |{"Key Share","For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                     
----------+-----+--------------------------
(0,1)     |t    |{"Key Share","For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_lcksvpt s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_lcksvpt: SELECT * FROM multixact_conflict FOR KEY SHARE; SAVEPOINT foo;
a
-
1
(1 row)

step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                     
----------+-----+--------------------------
(0,1)     |t    |{"Key Share","For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock1 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock1: SELECT * FROM multixact_conflict FOR KEY SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes            
----------+-----+-----------------
(0,1)     |f    |{"For Key Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes        
----------+-----+-------------
(0,1)     |f    |{"For Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes        
----------+-----+-------------
(0,1)     |f    |{"For Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes        
----------+-----+-------------
(0,1)     |f    |{"For Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock2 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock2: SELECT * FROM multixact_conflict FOR SHARE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes        
----------+-----+-------------
(0,1)     |f    |{"For Share"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                
----------+-----+---------------------
(0,1)     |f    |{"For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                
----------+-----+---------------------
(0,1)     |f    |{"For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                
----------+-----+---------------------
(0,1)     |f    |{"For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock3 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock3: SELECT * FROM multixact_conflict FOR NO KEY UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes                
----------+-----+---------------------
(0,1)     |f    |{"For No Key Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes         
----------+-----+--------------
(0,1)     |f    |{"For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes         
----------+-----+--------------
(0,1)     |f    |{"For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes         
----------+-----+--------------
(0,1)     |f    |{"For Update"}
(1 row)

step s1_commit: COMMIT;

starting permutation: s1_begin s1_tuplock4 s2_rowlocks s1_commit
step s1_begin: BEGIN;
step s1_tuplock4: SELECT * FROM multixact_conflict FOR UPDATE;
a
-
1
(1 row)

step s2_rowlocks: select locked_row, multi, modes from pgrowlocks('multixact_conflict');
locked_row|multi|modes         
----------+-----+--------------
(0,1)     |f    |{"For Update"}
(1 row)

step s1_commit: COMMIT;
